src/main/scala/firrtl/Visitor.scala

```diff
@@ -304,13 +300,22 @@ class Visitor(infoMode: InfoMode) extends FIRRTLBaseVisitor[FirrtlNode] {
           UIntLiteral(value, width)
         case "SInt" =>
           val (width, value) =
-            if (ctx.getChildCount > 4)
-              (IntWidth(string2BigInt(ctx.intLit(0).getText)), string2BigInt(ctx.intLit(1).getText))
-            else {
-              val bigint = string2BigInt(ctx.intLit(0).getText)
-              (IntWidth(BigInt(bigint.bitLength + 1)), bigint)
+            if (ctx.getChildCount > 4) {
+              val width = string2BigInt(ctx.intLit(0).getText)
+              val value = string2BigInt(ctx.intLit(1).getText)
+              (IntWidth(width), value)
+            } else {
+              val str = ctx.intLit(0).getText
+              val value = string2BigInt(str)
+              if (value < 0) {
+                val sintWidth = (value.abs - BigInt(1)).bitLength + 1
+                (IntWidth(BigInt(sintWidth)), value)
+              } else {
+                (IntWidth(BigInt(value.bitLength + 1)), value)
+              }
```

Comment:
I think this would better be expressed as
```scala
val width = if (value < 0) (value.abs - BigInt(1)).bitLength + 1 else value.bitLength + 1
(IntWidth(BigInt(width)), value)
```
Also would appreciate a comment explaining why the width is calculated the way it is